src/graph/planner/match/MatchClausePlanner.cpp

```diff
@@ -108,32 +116,54 @@ StatusOr<SubPlan> MatchClausePlanner::transform(CypherClauseContextBase* clauseC
   }
 
   auto* matchClauseCtx = static_cast<MatchClauseContext*>(clauseCtx);
-  auto& nodeInfos = matchClauseCtx->nodeInfos;
-  auto& edgeInfos = matchClauseCtx->edgeInfos;
   SubPlan matchClausePlan;
-  size_t startIndex = 0;
-  bool startFromEdge = false;
-
-  NG_RETURN_IF_ERROR(findStarts(matchClauseCtx, startFromEdge, startIndex, matchClausePlan));
-  NG_RETURN_IF_ERROR(
-      expand(nodeInfos, edgeInfos, matchClauseCtx, startFromEdge, startIndex, matchClausePlan));
+  // All nodes ever seen in current match clause
+  std::unordered_set<std::string> nodeAliasesSeen;
+  // TODO: Maybe it is better to rebuild the graph and find all connected components.
+  auto& pathInfos = matchClauseCtx->paths;
+  for (auto iter = pathInfos.begin(); iter < pathInfos.end(); ++iter) {
+    auto& nodeInfos = iter->nodeInfos;
+    auto& edgeInfos = iter->edgeInfos;
+    SubPlan subplan;
+    size_t startIndex = 0;
+    bool startFromEdge = false;
+
+    NG_RETURN_IF_ERROR(findStarts(
```

Comment:
Could we move the logic of `StartVidFinder` or `IndexSelection` to the optimizer in the future?
These logic makes the plannner complicated.

Reply:
I'm ok if you have a full solution, talk is always cheap.